pkg/collector/event.go

```diff
@@ -0,0 +1,180 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	coreV1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"github.com/caicloud/event_exporter/pkg/filters"
+	"github.com/caicloud/event_exporter/pkg/options"
+)
+
+type EventCollector struct {
+	kc                kubernetes.Interface
+	factory           informers.SharedInformerFactory
+	eventLister       coreV1.EventLister
+	eventListerSynced cache.InformerSynced
+	queue             workqueue.RateLimitingInterface
+	filters           []filters.EventFilter
+	cache             map[string]v1api.Event
+	locker            sync.Mutex
+}
+
+func NewEventCollector(kc kubernetes.Interface, factory informers.SharedInformerFactory, o *options.Options) *EventCollector {
+	event := factory.Core().V1().Events()
+	eventCollector := &EventCollector{
+		kc:                kc,
+		factory:           factory,
+		eventLister:       event.Lister(),
+		eventListerSynced: event.Informer().HasSynced,
+		queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		cache:             make(map[string]v1api.Event),
+		locker:            sync.Mutex{},
+	}
+	eventCollector.addFilter(o)
+	event.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			eventCollector.enqueueEvent(obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			newObj := new.(*v1api.Event)
+			oldObj := old.(*v1api.Event)
+			if newObj.ResourceVersion == oldObj.ResourceVersion {
+				return
+			}
+			eventCollector.enqueueEvent(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			eventCollector.enqueueEvent(obj)
+		},
+	})
+	return eventCollector
+}
+
+func (ec *EventCollector) Run(stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	klog.Info("starting eventCollector")
+	if ok := cache.WaitForCacheSync(stopCh, ec.eventListerSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+	klog.Info("started")
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+	klog.Info("shutting down")
+	return nil
+}
+
+func (ec *EventCollector) runWorker() {
+	for ec.processNextItem() {
+	}
+}
+
+func (ec *EventCollector) processNextItem() bool {
+	key, quit := ec.queue.Get()
+	if quit {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	err := ec.syncEvent(key.(string))
+	if err == nil {
+		ec.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("syncing %v failed with : %v", key, err))
+	ec.queue.AddRateLimited(key)
+	return true
+}
+
+func (ec *EventCollector) enqueueEvent(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("couldn't get key for object %+v:%v", obj, err))
+		return
+	}
+	ec.queue.Add(key)
+}
+
+func (ec *EventCollector) syncEvent(key string) error {
```

Comment:
comment required

Reply:
As coded in main.go, we set resync period to ensure the consistency of events.